src/app/CommandSender.h

```diff
@@ -124,8 +151,30 @@ class CommandSender final : public Messaging::ExchangeDelegate
     CommandSender(Callback * apCallback, Messaging::ExchangeManager * apExchangeMgr, bool aIsTimedRequest = false,
                   bool aSuppressResponse = false);
     ~CommandSender();
-    CHIP_ERROR PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct = true);
-    CHIP_ERROR FinishCommand(bool aEndDataStruct = true);
+
+    /**
+     * Enables CommandSender to send batch commands up until the number of supported by remote peer.
+     *
+     * It ensures that commands contain all required elements while building such that sending the
+     * InvokeRequestMessage has a high likeihood to succeed. Must be called before PrepareCommand.
+     * If aRemoteMaxPathsPerInvoke is 1, this will allow the call to contain only one command and
+     * doesn't enforce other batch commands requirements (similar to have never calling this method.
+     *
+     * @param [in] aRemoteMaxPathsPerInvoke max number of paths per invoke supported by remote peer.
+     *
+     * @return CHIP_ERROR_INCORRECT_STATE
+     *                      If device has previously called `PrepareCommand`.
+     * @return CHIP_ERROR_INVALID_ARGUMENT
+     *                      Invalid argument value.
+     * @return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE
+     *                      Device has not enabled CHIP_CONFIG_SENDING_BATCH_COMMANDS_ENABLED.
+     */
+    CHIP_ERROR SetBatchCommandsConfig(uint16_t aRemoteMaxPathsPerInvoke);
+
+    CHIP_ERROR PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct = true,
+                              AdditionalCommandDataElements aOptionalArgs = AdditionalCommandDataElements());
+    CHIP_ERROR FinishCommand(bool aEndDataStruct                         = true,
```

Comment:
bools as arguments to methods are often hard to read. In this case without reading arguments or documentation a `FinishCommand(true)` vs `FinishCommand(false)` is not obvious what they do.

Consider making these struct arguments (where we can use builders on) or using named enums for the commands , so I read `FinishCommand(Finish::EndDataStruct)` or `PrepareCommand(Prepare::StartDataStruct)` or `PrepareCommand(Prepare::kDoNotStartNewDataStruct)` (this may need some better reasonable names).

Reply:
I think you are altering these: you are adding a 3rd argument.

So I would propose to keep the old one with old arguments and make that an inline call to a `better version` instead of keeping extending to more arguments (which then become a stable API that we cannot touch anymore)

Reply:
Done, there is a little bit of weirdness since internally (in `AddRequestDataInternal` we want `mStartOrEndDataStruct` to be false, but for all calls outside of SDK we want the default to be true.

But I think I dealt with it in the most sane way